src/main/java/org/ice4j/util/AsyncQueueHandler.java

```diff
@@ -0,0 +1,261 @@
+     * same {@link ExecutorService}.
+     */
+    private final long maxSequentiallyHandledItems;
+
+    /**
+     * A flag which indicates if reading of {@link #queue} is allowed
+     * to continue.
+     */
+    private final AtomicBoolean running = new AtomicBoolean();
+
+    /**
+     * Synchronization object of current instance state, in particular
+     * used to resolve races between {@link #handleQueueItemsUntilEmpty()}
+     * and {@link #reader} exit. In particular synchronization object used to
+     * access to field {@link #readerFuture}.
+     */
+    private final Object syncRoot = new Object();
+
+    /**
+     * Stores <tt>Future</tt> of currently executing {@link #reader}
+     */
+    private Future<?> readerFuture;
+
+    /**
+     * Perpetually reads item from {@link #queue} and uses
+     * {@link #handler} on each of them.
+     */
+    private final Runnable reader = new Runnable()
+    {
+        @Override
+        public void run()
+        {
+            long sequentiallyHandledItems = 0;
+
+            while (running.get())
+            {
+                if (maxSequentiallyHandledItems > 0 &&
+                    sequentiallyHandledItems >= maxSequentiallyHandledItems)
+                {
+                    onYield();
+                    return;
+                }
+
+                T item;
+
+                synchronized (syncRoot)
+                {
+                    item = queue.poll();
+
+                    if (item == null)
+                    {
+                        cancel(false);
+                        return;
+                    }
+                }
+
+                sequentiallyHandledItems++;
+
+                try
+                {
+                    handler.handleItem(item);
+                }
+                catch (Throwable e)
+                {
+                    logger.warning("Failed to handle item: " + e);
+                }
+            }
+        }
+    };
+
+    /**
+     * Constucts instance of {@link AsyncQueueHandler<T>} which is capable of
+     * asyncronous reading provided queue from thread borrowed from executor to
+     * process items with provided handler.
+     * @param queue thread-safe queue which holds items to process
+     * @param handler an implementation of handler routine which will be
+     * invoked per each item placed in the queue.
+     * @param id optional identifier of current handler for debug purpose
+     * @param executor optional executor service to borrow threads from
+     * @param maxSequentiallyHandledItems maximum number of items sequentially
+     * handled on thread borrowed from {@link #executor} before temporary
+     * releasing thread and re-acquiring it from {@link #executor}.
+     */
+    public AsyncQueueHandler(
+        BlockingQueue<T> queue,
+        Handler<T> handler,
+        String id,
+        ExecutorService executor,
+        long maxSequentiallyHandledItems)
```

Comment:
nit: can we add another ctor that doesn't take in a `maxSequentiallyHandledItems` and defaults it to no limit?